operator/pkg/ciliumenvoyconfig/manager.go

```diff
@@ -0,0 +1,86 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright 2019-2020 Authors of Cilium
+
+package ciliumenvoyconfig
+
+import (
+	"context"
+
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	slim_corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	"github.com/cilium/cilium/pkg/k8s/watchers/subscriber"
+)
+
+var _ subscriber.Service = &manager{}
+
+const (
+	ciliumEnvoyLBPrefix = "cilium-envoy-lb"
+)
+
+type manager struct {
+	envoyConfigManager *envoyConfigManager
+
+	queue      workqueue.RateLimitingInterface
+	maxRetries int
+
+	serviceStore cache.Store
+}
+
+// New returns a new manager for CiliumEnvoyConfig
+func New(ctx context.Context, indexer cache.Store) (*manager, error) {
+	manager := &manager{
+		queue:        workqueue.NewRateLimitingQueue(workqueue.DefaultControllerRateLimiter()),
+		serviceStore: indexer,
+		maxRetries:   10,
+	}
+
+	envoyConfigManager, err := newEnvoyConfigManager(manager.maxRetries)
+	if err != nil {
+		return nil, err
+	}
+	manager.envoyConfigManager = envoyConfigManager
+
+	return manager, nil
+}
+
+func (m *manager) OnAddService(service *slim_corev1.Service) error {
+
+	return nil
+}
+
+func (m *manager) OnUpdateService(oldObj, newObj *slim_corev1.Service) error {
+	return nil
+}
+
+func (m *manager) OnDeleteService(service *slim_corev1.Service) error {
+	return nil
+}
+
+// Run kicks off the controlled loop
+func (m *manager) Run() {
+	defer m.queue.ShutDown()
+	for {
+		ev, quit := m.queue.Get()
+		if quit {
+			return
+		}
+		err := m.processEvent(ev)
+		if err != nil {
+			if m.queue.NumRequeues(ev) < m.maxRetries {
```

Comment:
this snippet seems to get repeated a lot in cilium, wonder if we should factor this out someday.

Reply:
yup, I just came to know the new generic pattern added recently. I think after sometime, we can adopt it widely.